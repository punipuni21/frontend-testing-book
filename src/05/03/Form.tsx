type Props = {
  name: string;
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void; // whats's this?
};

export const Form = ({ name, onSubmit }: Props) => {
  return (
    <form
      onSubmit={(event) => {
        event.preventDefault();
        onSubmit?.(event);
      }}
    >
      <h2>account info</h2>
      <p>{name}</p>
      <div>
        <button>edit</button>
      </div>
    </form>
  );
};
