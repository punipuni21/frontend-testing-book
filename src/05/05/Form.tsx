import { useId, useState } from "react";
import { InputAccount } from "./InputAccount";
import { Agreement } from "./Agreement";

export const Form = () => {
  const [checked, setChecked] = useState(false);
  const headingId = useId();

  return (
    <form area-labelledby={headingId}>
      <h2 id={headingId}>register new account</h2>
      <InputAccount />
      <Agreement
        onChange={(event) => setChecked(event.currentTarget.checked)}
      />
      <div>
        <button disabled={!checked}>sign up</button>
      </div>
    </form>
  );
};
