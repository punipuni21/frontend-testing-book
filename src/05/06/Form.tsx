import React, { useState } from "react";
import { ContactNumber } from "./ContactNumber";
import { RegisterDeliveryAddress } from "./RegisterDeliverryAddress";
import { DeliveryAddress } from "./DeliverryAddress";
import { PastDeliveryAddress } from "./PastDeliverryAddress";

export type AddressOption = React.ComponentProps<"option"> & { id: string };
export type Props = {
  deliveryAddress: AddressOption[];
  onSubmit?: (event: React.FormEvent<HTMLFormElement>) => void;
};

export const Form = (props: Props) => {
  const [registerNew, setRegisterNew] = useState<boolean | undefined>(
    undefined
  );
  return (
    <form onSubmit={props.onSubmit}>
      <h2>お届け先情報の入力</h2>
      <ContactNumber />
      {props.deliveryAddress?.length ? (
        <>
          <RegisterDeliveryAddress onChange={setRegisterNew} />
          {registerNew ? (
            <DeliveryAddress title="新しいお届け先" />
          ) : (
            <PastDeliveryAddress
              disabled={registerNew === undefined}
              options={props.deliveryAddress}
            />
          )}
        </>
      ) : (
        <DeliveryAddress />
      )}
      <hr />
      <div>
        <button>注文内容の確認へ進む</button>
      </div>
    </form>
  );
};
