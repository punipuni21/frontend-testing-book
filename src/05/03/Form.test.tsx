import { fireEvent, render, screen } from "@testing-library/react";
import { Form } from "./Form";

// @babel/preset-typescript: TypeScriptからJavaScriptへ変換
// @babel/preset-react: BabelにJSX→JSへ変換してもらう
// @babel/preset-env : ESModules→CommonJSの変換

test("display name", () => {
  render(<Form name="taro" />);
  expect(screen.getByText("taro")).toBeInTheDocument();
});

test("display button", () => {
  render(<Form name="taro" />);
  expect(screen.getByRole("button")).toBeInTheDocument();
});

test("display heading", () => {
  render(<Form name="taro" />);
  expect(screen.getByRole("heading")).toHaveTextContent("account info");
});

test("when push button, then event handler is called", () => {
  const mockFn = jest.fn();
  render(<Form name="taro" onSubmit={mockFn} />);
  fireEvent.click(screen.getByRole("button"));
  expect(mockFn).toHaveBeenCalled();
});
