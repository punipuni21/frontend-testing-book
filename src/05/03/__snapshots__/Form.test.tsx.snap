// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot: display account name 'taro' 1`] = `
<div>
  <form>
    <h2>
      account info
    </h2>
    <p>
      taro
    </p>
    <div>
      <button>
        edit
      </button>
    </div>
  </form>
</div>
`;
